### PR TITLE
Implement offset for crime data

### PR DESCRIPTION
**Pull request by s3364859**
@6nop @s3364942 @s3494110 @s3482247 

In this PR:
- Implemented offset for crime data, which required some fiddling about
- Added in the ability to export crime data, but commented it out as it's not very nice. Could use an extra set of eyes here
- Please do not add code to this PR

crime

### CODE REVIEW

File: CPT331.Core/ApplicationConfiguration.cs [hunk 1]
```diff
@@ -46,6 +48,30 @@ public string CPT331ConnectionString
 		}
 
 		/// <summary>
+		/// Gets the size of the number of rows to return.
+		/// </summary>
+		[ApplicationScopedSetting]
+		[DefaultSettingValue(DefaultDataTakeSize)]
+		public int DataTakeSize
```
Comment (s3364859) [inline CPT331.Core/ApplicationConfiguration.cs]: Number of records to select where OFFSET is implemented.

File: CPT331.Core/ObjectModel/CrimeOffenceLocalGovernmentAreaState.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+#endregion
+
+namespace CPT331.Core.ObjectModel
+{
+	/// <summary>
+	/// Represents a CrimeOffenceLocalGovernmentAreaState type, used to describe information about a crime, offence, local government area, and state or territory.
+	/// </summary>
+	public class CrimeOffenceLocalGovernmentAreaState : Crime
+	{
```
Comment (s3364859) [inline CPT331.Core/ObjectModel/CrimeOffenceLocalGovernmentAreaState.cs]: Represents the new view which joins this information together.

File: CPT331.Data.Migration/Migrations/Crime.CrimeOffenceLocalGovernmentAreaState.sql [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿CREATE VIEW Crime.CrimeOffenceLocalGovernmentAreaState
```
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/Crime.CrimeOffenceLocalGovernmentAreaState.sql]: New view for the crime data.

File: CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql [hunk 4]
```diff
@@ -0,0 +1,61 @@
+﻿CREATE PROCEDURE Crime.spGetCrimeOffenceLocalGovernmentAreaState
+
+@Skip INT = 0,
+@Take INT = 100,
+@OrderBy VARCHAR(10) = 'ID',
+@SortDirection VARCHAR(4) = 'ASC'
+
+AS
+
+IF (@OrderBy IS NULL) OR (@OrderBy = '')
```
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql]: Not sure if these are needed?

File: CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+BEGIN
+	SET @OrderBy = 'ID'
+END
+
+IF (@SortDirection IS NULL) OR (@SortDirection = '')
```
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql]: Or this one either?

File: CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	IsVisible,
+	DateCreatedUtc,
+	DateUpdatedUtc
+FROM
+	Crime.CrimeOffenceLocalGovernmentAreaState
+
+ORDER BY
+	ID
+	/*
+	CASE WHEN @OrderBy = 'Count' AND @SortDirection = 'ASC' THEN [Count] END ASC,
```
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql]: Regardless, I spent a little bit of time getting this to work, and it wasn't very quick, so I commented it out and though that perhaps someone smarter than me can fix it in a later PR. I'm looking at you, @6nop!

File: CPT331.Data/CrimeOffenceLocalGovernmentAreaStateRepository.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+using CPT331.Core.ObjectModel;
+
+#endregion
+
+namespace CPT331.Data
+{
+	/// <summary>
+	/// Represents a CrimeOffenceLocalGovernmentAreaStateRepository type, used to manipulate crime, offence, local government area, and state or territory data.
+	/// </summary>
+	public class CrimeOffenceLocalGovernmentAreaStateRepository : Repository
```
Comment (s3364859) [inline CPT331.Data/CrimeOffenceLocalGovernmentAreaStateRepository.cs]: New repository type.

File: CPT331.Data/CrimeRepository.cs [hunk 8]
```diff
@@ -115,18 +129,36 @@ public List<CrimeByCoordinate> GetCrimesByCoordinate(double latitude, double lon
...
+		/// <returns>Returns a list of Crime objects representing the result of the operation.</returns>
+		public List<Crime> GetCrimes(int skip, int take)
+		{
 			List<Crime> crimes = null;
 
 			using (SqlConnection sqlConnection = SqlConnectionFactory.NewSqlConnetion())
 			{
 				crimes = SqlMapper
-					.Query(sqlConnection, CrimeSpGetCrime, commandType: CommandType.StoredProcedure)
+					.Query(sqlConnection, CrimeSpGetCrime, new { Skip = skip, Take = take }, commandType: CommandType.StoredProcedure)
```
Comment (s3364859) [inline CPT331.Data/CrimeRepository.cs]: Add in paging bits.

File: CPT331.Data/CrimeRepository.cs [hunk 9]
```diff
@@ -139,6 +171,32 @@ public List<Crime> GetCrimes()
 
 			return crimes;
 		}
+		
+		/// <summary>
+		/// Exports a list of ReadOnlyDataObject types to a stream.
+		/// </summary>
+		/// <param name="readOnlyDataObjects">The list of ReadOnlyDataObject objects to export.</param>
+		/// <param name="stream">The stream to export the ReadOnlyDataObject list to.</param>
+		protected override void OnExport(List<ReadOnlyDataObject> readOnlyDataObjects, Stream stream)
```
Comment (s3364859) [inline CPT331.Data/CrimeRepository.cs]: So this doesn't work quite the way I want it to yet - ideally it should stream from the database to the browser - which it doesn't. Hopefully can be tweaked in a later PR.

File: CPT331.Web/Views/Admin/Crime.cshtml [hunk 10]
```diff
@@ -0,0 +1,78 @@
+﻿@using CPT331.Web.Models.Admin
```
Comment (s3364859) [inline CPT331.Web/Views/Admin/Crime.cshtml]: New views for crime data.

File: CPT331.Data/DataProvider.cs [hunk 11]
```diff
@@ -173,13 +128,17 @@ public static UserRepository UserRepository
...
-					_userRepository = new UserRepository();
-				}
-
-				return _userRepository;
-			}
+                return GetDataRepository(_userRepository);
+            }
 		}
+
+        private static T GetDataRepository<T>(T instanceVariable) where T : Repository, new()
```
Comment (6nop) [inline CPT331.Data/DataProvider.cs]: Generic implementation of getter code

File: CPT331.Data/DataProvider.cs [hunk 12]
```diff
@@ -29,12 +29,7 @@ public static AdhocScriptRepository AdhocScriptRepository
 		{
 			get
 			{
-				if (_adhocScriptRepository == null)
-				{
-					_adhocScriptRepository = new AdhocScriptRepository();
-				}
-
-				return _adhocScriptRepository;
+				return GetDataRepository(_adhocScriptRepository);
```
Comment (6nop) [inline CPT331.Data/DataProvider.cs]: Moved duplicated implementation to a private generic method

File: CPT331.Data.Migration/Migrations/Crime.spGetCrime.sql [hunk 13]
```diff
@@ -11,7 +11,7 @@ SELECT
 	OffenceID,
 	[Count],
 	[Month],
-	[Year]
+	[Year],
 	IsDeleted,
```
Comment (6nop) [inline CPT331.Data.Migration/Migrations/Crime.spGetCrime.sql]: Fixed missing comma between ```Year``` and ```IsDeleted```

File: CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql (hunk 5 above)
Comment (6nop) [inline CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql]: Based on our current code (```GetCrimeOffenceLocalGovernmentAreaStates()```), I believe it is possible to hit this SP with blanks (_OrderBy only_) or nulls (_OrderBy & SortDirection_).

So it's probably safer to leave this code in there.

File: CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql (hunk 6 above)
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/Crime.spGetCrimeOffenceLocalGovernmentAreaState.sql]: Yep, probably too hard for this commit.

File: CPT331.Data.Migration/Migrations/Crime.spGetCrime.sql (hunk 13 above)
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/Crime.spGetCrime.sql]: shifty-eyes-o

**Pull request merged**